### PR TITLE
Merge 2.8

### PR DESCRIPTION
**Pull request by wallyworld**
## Description of change

Merge 2.8 with these PRs:

#11854 Handle empty API addresses
#11859 centos7-lxd
#11865 Add a juju repl CLI
#11866 Improve juju version help text
#11853 Use model operation for updating machine sourced link layer data
#11864 Improve openstack network config error messages
#11868 Fix support for juju -version

## QA steps

See PRs

### CODE REVIEW

**Comment by manadart** [comment]
Looks like there's a regression here.
```
FAIL: main_test.go:47: k8sAgentSuite.TestRegisteredSubCommandsForK8sAgentCommand

main_test.go:54:
    c.Assert(cmdtesting.Stdout(ctx), gc.Equals, `
    help     Show help on a command or other topic.
    init     initialize k8sagent state
    unit     starting a k8s agent
    `[1:])
... obtained string = "" +
...     "help  Show help on a command or other topic.\n" +
...     "init  initialize k8sagent state\n" +
...     "unit  starting a k8s agent\n"
... expected string = "" +
...     "help     Show help on a command or other topic.\n" +
...     "init     initialize k8sagent state\n" +
...     "unit     starting a k8s agent\n"
... String difference:
...     [0]: "help  Show help on a command or other topic." != "help     Show help on a command or other topic."
...     [1]: "init  initialize k8sagent state" != "init     initialize k8sagent state"
...     [2]: "unit  starting a k8s agent" != "unit     starting a k8s agent"
```

I think we need to get on top of more frequent forward merges. This one has one of mine that landed late yesterday, but the last few times I've done it there have been ~40 file changes; it gets hard to review properly.

**Comment by wallyworld** [comment]
@manadart thanks for view, fixed test failure, just had to delete a few spaces from some expected text

File: apiserver/common/networkingcommon/package_test.go [hunk 1]
```diff
@@ -1,16 +1,43 @@
...
 
-package networkingcommon_test
+package networkingcommon
 
 import (
-	stdtesting "testing"
+	"testing"
 
-	coretesting "github.com/juju/juju/testing"
+	"github.com/juju/juju/state"
```
Comment (manadart) [inline apiserver/common/networkingcommon/package_test.go:9]: Missed this grouping in my patch...

Comment (wallyworld) [reply apiserver/common/networkingcommon/package_test.go]: doh, thanks

**Comment by wallyworld** [comment]
$$merge$$

**Pull request merged**